File: services/sttm-builder/app/schema/dbt_conversion.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

from app.schema.common import TableRef
from app.schema.mapping_sql import MappingSqlMappingItem
from app.schema.sttm_builder import RelationshipContextItem, SemanticContextItem


class DbtConversionRequest(BaseModel):
    project_name: str | None = None
    domain_name: str | None = None
    target_layer: Literal["raw", "curated", "mart"] | str | None = None
    materialization: Literal["incremental", "table", "view"] | str | None = None
    source_tables: list[TableRef] = Field(default_factory=list)
    target_table: TableRef
    driving_table: TableRef | None = None
    selected_derived_sources: list[str] = Field(default_factory=list)
    relationships: list[RelationshipContextItem] = Field(default_factory=list)
    selected_columns_by_table: dict[str, list[str]] = Field(default_factory=dict)
    semantic_bundle_id: str | None = None
    semantic_bundle_label: str | None = None
    semantic_view_name: str | None = None
    source_query_sql: str | None = None
    validated_sql: str = Field(min_length=1)
    generated_sql: str | None = None
    mappings: list[MappingSqlMappingItem] = Field(default_factory=list)
    semantic_context: list[SemanticContextItem] = Field(default_factory=list)
    derived_source_lineage: list[dict[str, Any]] = Field(default_factory=list)
    datahub_context: dict[str, Any] | None = None
    checklist: list[str] = Field(default_factory=list)


class DbtConversionGeneratedFile(BaseModel):
    file_name: str
    file_path: str
    file_type: str = "MODEL"
    content: str
    language: Literal["sql", "yaml", "text"] = "sql"


class DbtConversionSourceUpdate(BaseModel):
    file_path: str
    action: Literal["UPDATE", "NO_CHANGE"] | str
    content: str | None = None
    language: Literal["yaml", "text"] = "yaml"


class DbtConversionResponse(BaseModel):
    status: Literal["completed", "failed"] | str
    action: Literal["CREATE_NEW", "UPDATE_EXISTING", "CREATE_WITH_REFERENCE", "REUSE_EXISTING"] | str | None = None
    message: str | None = None
    generated_files: list[DbtConversionGeneratedFile] = Field(default_factory=list)
    source_update: DbtConversionSourceUpdate | None = None
    schema_files: list[DbtConversionGeneratedFile] = Field(default_factory=list)
    macros_used: list[str] = Field(default_factory=list)
    materialization: str | None = None
    materialization_reason: str | None = None
    agent_name: str
    domain_name: str | None = None
    target_layer: str | None = None
    branch: str = "main"

File: frontend/src/data/mock/dbt-conversion.ts
import { buildApiEnvelope, createRequestId } from "@/api/axiosInstance";
import type {
  DbtConversionGeneratedFile,
  DbtConversionRequest,
  DbtConversionResponse,
} from "@/types/api-contract";
import { mockSleep } from "@/services/mock/mockConfig";

function slugify(value: string) {
  return value.toLowerCase().replace(/[^a-z0-9]+/g, "_").replace(/^_|_$/g, "");
}

function buildMockDbtFiles(payload: DbtConversionRequest): {
  generatedFiles: DbtConversionGeneratedFile[];
  schemaFiles: DbtConversionGeneratedFile[];
} {
  const modelName = slugify(payload.target_table.table);
  const domain = slugify(payload.domain_name ?? payload.project_name ?? "sales_core");
  const sql = payload.validated_sql?.trim() || payload.generated_sql?.trim() || "SELECT 1";

  const generatedFiles: DbtConversionGeneratedFile[] = [
    {
      file_name: `${modelName}.sql`,
      file_path: `models/${domain}/${modelName}.sql`,
      file_type: "model",
      language: "sql",
      content: [
        "-- Mock dbt model generated for local development",
        `{{ config(materialized='${payload.materialization ?? "incremental"}') }}`,
        "",
        sql,
      ].join("\n"),
    },
  ];

  const schemaFiles: DbtConversionGeneratedFile[] = [
    {
      file_name: `${modelName}.yml`,
      file_path: `models/${domain}/${modelName}.yml`,
      file_type: "schema",
      language: "yaml",
      content: [
        "version: 2",
        "models:",
        `  - name: ${modelName}`,
        `    description: Mock schema for ${payload.target_table.table}`,
        "    columns:",
        ...payload.mappings.slice(0, 8).map((mapping) => [
          `      - name: ${mapping.target_column}`,
          `        description: Mock column mapping for ${mapping.target_column}`,
        ].join("\n")),
      ].join("\n"),
    },
  ];

  return { generatedFiles, schemaFiles };
}

export function buildMockDbtConversion(
  payload: DbtConversionRequest,
): DbtConversionResponse {
  const { generatedFiles, schemaFiles } = buildMockDbtFiles(payload);

  return {
    status: "completed",
    action: "CREATE_NEW",
    message: "Mock dbt conversion completed using local development data.",
    generated_files: generatedFiles,
    schema_files: schemaFiles,
    source_update: {
      file_path: "models/sources.yml",
      action: "NO_CHANGE",
      content: null,
      language: "yaml",
    },
    macros_used: ["mock_generate_surrogate_key"],
    materialization: payload.materialization ?? "incremental",
    materialization_reason: "Mock materialization selected for local preview.",
    agent_name: "MOCK_DBT_CONVERSION_AGENT",
    domain_name: payload.domain_name ?? domainFromPayload(payload),
    target_layer: payload.target_layer ?? "curated",
    branch: "feature/mock-dbt-conversion",
  };
}

function domainFromPayload(payload: DbtConversionRequest) {
  return slugify(payload.domain_name ?? payload.target_table.schema ?? "sales_core");
}

function buildMockDbtFinalEnvelope(payload: DbtConversionRequest) {
  const result = buildMockDbtConversion(payload);
  return buildApiEnvelope("dbt_conversion.generate", result, {
    target_table: payload.target_table,
    source_tables: payload.source_tables,
    driving_table: payload.driving_table ?? null,
    relationships: payload.relationships ?? [],
    semantic_bundle_id: payload.semantic_bundle_id ?? null,
    semantic_view_name: payload.semantic_view_name ?? null,
    thread_id: `mock-dbt-${createRequestId().slice(0, 8)}`,
  });
}

export async function* mockStreamDbtConversion(
  payload: DbtConversionRequest,
): AsyncGenerator<
  | { event: "status"; data: Record<string, unknown> }
  | { event: "artifact"; data: Record<string, unknown> }
  | { event: "final"; data: Record<string, unknown> }
  | { event: "error"; data: { message?: string; code?: string } }
> {
  yield {
    event: "status",
    data: {
      phase: "agent_progress",
      message: "AGT_DBT_CONVERSION is reviewing the final SQL and loading dbt project context.",
    },
  };
  await mockSleep(250);

  const { generatedFiles, schemaFiles } = buildMockDbtFiles(payload);

  for (const file of generatedFiles) {
    yield {
      event: "status",
      data: {
        phase: "file_ready",
        message: `Generated ${file.file_path}.`,
      },
    };
    yield {
      event: "artifact",
      data: {
        kind: "generated_file",
        file,
      },
    };
    await mockSleep(120);
  }

  for (const file of schemaFiles) {
    yield {
      event: "status",
      data: {
        phase: "file_ready",
        message: `Prepared ${file.file_path}.`,
      },
    };
    yield {
      event: "artifact",
      data: {
        kind: "schema_file",
        file,
      },
    };
    await mockSleep(120);
  }

  const finalEnvelope = buildMockDbtFinalEnvelope(payload);
  yield {
    event: "final",
    data: finalEnvelope as unknown as Record<string, unknown>,
  };
}
